### PR TITLE
Event Registration Modal

### LINKED ISSUE #57: When "register" button is clicked prompt the user to login
[Project/Issue Tracker](https://github.com/Code-4-Community/lucys-love-bus-frontend/projects/1)

### Objective

As a user, I want to be prompted to login if I click on "register" in the single events view if I am not already logged in.

However, if this is too difficult then we can simply display a tooltip and disable the button
 
### Tech Spec

- [ ] When the user clicks on the "Register" button in the single events view and they are not logged in, open the login modal. OR if not possible then disable the button and show a tooltip prompting the user to log in

### PR DESCRIPTION
**Pull request by ja0911**
## Issue

Closes #57, #68

## Changes

- [X] When the user clicks on the "Register" button in the single events view and they are not logged in, open the login modal. OR if not possible then disable the button and show a tooltip prompting the user to log in
- [X] Clicking on the "register" button on the single event page should open up a modal
   - [X] Containing a small form with an input for how many tickets to purchase, and a submit button 
   - [X] Also make it possible to cancel the registration modal
- [X] Registrations for events for participating families and general members  will currently be free of charge, no need to use stripe at all, just make the appropriate API request to the backend.

## Screenshots

image

image

## Verification Steps

- Manually tested logging in and registering for an event for different privilege levels
- Verifying that informative errors are displayed correctly to the user

### CODE REVIEW

File: src/store.ts [hunk 1]
```diff
@@ -136,6 +136,7 @@ const store: Store<C4CState, C4CAction> = createStore<
   {}
 >(reducers, preloadedState || initialStoreState, enhancer);
 
+/*
```
Comment (rymaju) [inline src/store.ts:139]: ‼️  This should be uncommented!

File: src/components/event-details/EventDetails.tsx [hunk 2]
```diff
@@ -85,7 +94,7 @@ const EventListing: React.FC<EventProps> = ({ thumbnail, title, details }) => {
       <TopRow>
         <CardContent>
           <Col span={14}>
-            <Thumbnail src={thumbnail || defaultImg}></Thumbnail>
+            <Thumbnail src={thumbnail || DEFAULT_IMAGE}></Thumbnail>
```
Comment (rymaju) [inline src/components/event-details/EventDetails.tsx:97]: 👍

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    if (typeof newValue === 'number') {
+      setQuantity(newValue);
+    } else {
+      setQuantity(1);
+    }
+  };
+
+  const handleOk = async () => {
+    try {
+      await protectedApiClient.registerTickets({
```
Comment (rymaju) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:76]: ⛏️  We may want to use an AsyncRequest for this. @jackblanc

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    } catch (e) {
+      setErrorMessage(e.response.data);
+    }
+  };
+  return (
+    <div>
+      <StyledModal
+        visible={showEventRegistrationModal}
+        title={eventTitle}
+        okButtonProps={{ disabled: privilegeLevel === PrivilegeLevel.NONE }}
```
Comment (rymaju) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:95]: ⛏️  We should also keep track if the current request is loading or not and disable the button if it is loading. We dont want users to accidentally send multiple requests and error!

**Review by rymaju — CHANGES_REQUESTED** [review]
You need to uncomment the throttle code! Thats essential for our localstorage redux store to work!

Fixing that will fix CI as well. Other than that I think it looks pretty good! I left some small nitpicks ⛏️ .

**Comment by beccajohnson17** [comment]
can i get a mini demo of this tonight?

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx (hunk 4 above)
Comment (vrautela) [reply src/components/modals/event-registration-modal/EventRegistrationModal.tsx]: I just added a super basic loading check. We can dump it if we switch to using an async request

File: src/api/protectedApiClient.ts [hunk 5]
```diff
@@ -11,6 +12,15 @@ export interface ProtectedApiClient {
...
   }) => Promise<void>;
+
+  readonly registerTickets: (request: {
+    lineItemRequests: [
+      {
+        eventId: number;
+        quantity: number;
+      },
+    ];
+  }) => Promise<AxiosResponse<any>>;
```
Comment (jackblanc) [inline src/api/protectedApiClient.ts:23]: Is there an interface for this request anywhere? if so we should reference it here. If not we should define one.

File: src/api/protectedApiClient.ts [hunk 6]
```diff
@@ -1,4 +1,5 @@
 import AppAxiosInstance from '../auth/axios';
+import { AxiosResponse } from 'axios';
```
Comment (jackblanc) [inline src/api/protectedApiClient.ts:2]:
```suggestion
```

File: src/api/protectedApiClient.ts (hunk 5 above)
Comment (jackblanc) [inline src/api/protectedApiClient.ts:23]: We should return Promise<void> here if there's no response we care about.

File: src/api/protectedApiClient.ts [hunk 7]
```diff
@@ -37,6 +48,19 @@ const changePassword = (request: {
...
 };
 
+const registerTickets = (request: {
+  lineItemRequests: [
+    {
+      eventId: number;
+      quantity: number;
+    },
+  ];
+}) => {
```
Comment (jackblanc) [inline src/api/protectedApiClient.ts:58]: interface here

File: src/api/protectedApiClient.ts [hunk 8]
```diff
@@ -50,9 +74,9 @@ const getRequestStatuses = (): Promise<PersonalRequest[]> => {
 };
 
 const makePFRequest = (): Promise<void> => {
-  return AppAxiosInstance.post(ProtectedApiClientRoutes.MAKE_PF_REQUEST)
-    .then(() => {})
-    .catch((err) => err);
+  return AppAxiosInstance.post(ProtectedApiClientRoutes.MAKE_PF_REQUEST).catch(
+    (err) => err,
+  );
```
Comment (jackblanc) [inline src/api/protectedApiClient.ts:79]: Why this change?

File: src/components/event-details/EventDetails.tsx [hunk 9]
```diff
@@ -57,9 +59,16 @@ const AnnouncementBox = styled.div`
...
-    'https://lucys-love-bus-public.s3.us-east-2.amazonaws.com/LLB_2019_Sq_rgb+1.png';
+const EventListing: React.FC<EventProps> = ({
+  thumbnail,
+  id,
+  title,
+  details,
+}) => {
+  const [
+    displayEventRegistrationModal,
+    setDisplayEventRegistrationModal,
```
Comment (jackblanc) [inline src/components/event-details/EventDetails.tsx:70]: ⛏️ a bit long of a name IMO

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+  width: 100%;
+`;
+const LeftAlignedText = styled(Text)`
+  display: block;
+  text-align: left;
+`;
+const BoldCenterText = styled(Text)`
+  font-size: 16px;
+  font-weight: 600;
+`;
```
Comment (jackblanc) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:45]: We should have one line of space between each of these

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx (hunk 3 above)
Comment (jackblanc) [reply src/components/modals/event-registration-modal/EventRegistrationModal.tsx:78]: Yeah, will explain in another comment.

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  onCloseEventRegistrationModal,
+  showEventRegistrationModal,
+  eventId,
+  eventTitle,
+}) => {
+  const [quantity, setQuantity] = React.useState<number>(1);
+  const [errorMessage, setErrorMessage] = React.useState<string | undefined>(
+    undefined,
+  );
+  const [loading, setLoading] = React.useState<boolean>(false);
```
Comment (jackblanc) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:64]: We can use AsyncRequest here:
`const [registrationRequest, setRegistrationRequest] = useState<AsyncRequest<void, any>>(AsyncRequestNotStarted())`

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+  const privilegeLevel: PrivilegeLevel = getPrivilegeLevel(tokens);
```
Comment (jackblanc) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:65]: I would prefer we pass privilegeLevel in directly, instead of the token AsyncRequest

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+  const updateQuantity = (newValue: string | number | undefined) => {
+    if (typeof newValue === 'number') {
+      setQuantity(newValue);
+    } else {
+      setQuantity(1);
+    }
+  };
```
Comment (jackblanc) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:73]: When would it be a string? why do we ignore strings?

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        ],
+      });
+      onCloseEventRegistrationModal();
+      setErrorMessage(undefined);
+      setLoading(false);
+    } catch (e) {
+      setErrorMessage(e.response.data);
+      setLoading(false);
+    }
+  };
```
Comment (jackblanc) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:93]: With the AsyncRequest useState above, this will change to look like: 
```suggestion
  const handleOk = async () => {
    try {
      setRegistrationRequest(AsyncRequestIsLoading());
      await protectedApiClient.registerTickets({
        lineItemRequests: [
          {
            eventId,
            quantity,
          },
        ],
      });
      setRegistrationRequest(AsyncRequestIsComplete());
      onCloseEventRegistrationModal();
    } catch (e) {
      setRegistrationRequest(AsyncRequestFailed(e));
    }
  };
```

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+        title={eventTitle}
+        okButtonProps={{
+          disabled: loading || privilegeLevel === PrivilegeLevel.NONE,
+        }}
+        onOk={handleOk}
+        okText={'Register'}
+        onCancel={() => {
+          onCloseEventRegistrationModal();
+        }}
+        width={'625px'}
```
Comment (jackblanc) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:107]:
```suggestion
        width="625px"
```

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+      >
+        {errorMessage !== undefined && (
+          <AlertWithMargin type="error" message={errorMessage} />
+        )}
+        <ContentDiv>
+          {privilegeLevel === PrivilegeLevel.NONE ? (
+            <BoldCenterText>
+              Please log in to register for this event.
+            </BoldCenterText>
```
Comment (jackblanc) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:116]: Can we make this be clickable/open the login modal (potentially "above" this one, so they can log in and then this modal is still open, at which point they can complete the registration?)

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+  );
+};
+
+const mapStateToProps = (state: C4CState): StateProps => {
+  return {
+    tokens: state.authenticationState.tokens,
+  };
+};
+
+export default connect(mapStateToProps)(EventRegistrationModal);
```
Comment (jackblanc) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:142]: So here, you're actually passing tokens twice: 
- when you call your component, you provide tokens as a prop
- this connect function will _also_ provide the same tokens as a prop
Not sure which comes first, but the second just overwrites. 

I think what we should do is have this component just take PrivilegeLevel directly. 

Also generally, only containers should be directly connected to the store, not components.

**Review by jackblanc — COMMENTED** [review]
Very excited this is coming together! Left some suggestions, let me know if you want to pair on anything.

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx (hunk 16 above)
Comment (rymaju) [reply src/components/modals/event-registration-modal/EventRegistrationModal.tsx:116]: The login modal state is currently local to the navbar, so this wouldnt be possible as of now.

File: src/api/protectedApiClient.ts (hunk 8 above)
Comment (vrautela) [reply src/api/protectedApiClient.ts:79]: image

We didn't add these lines, we just got rid of the .then because it was complaining. Totally down to change this to whatever it should be

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx (hunk 13 above)
Comment (vrautela) [reply src/components/modals/event-registration-modal/EventRegistrationModal.tsx:73]: It would be a string if the user entered in something like "fred". In most cases it would just be a number or undefined. Here we ignore all of the inputs that aren't numbers by setting the value in the input field to be 1

**Review by rymaju — APPROVED** [review]
LGTM 👍

**Review by jackblanc — APPROVED** [review]
Nice work :)

File: src/components/event-details/EventDetails.tsx [hunk 18]
```diff
@@ -1,11 +1,26 @@
...
+  thumbnail?: string;
+  details: {
+    description: string;
+    location: string;
+    start: Date;
+    end: Date;
+  };
+  privilegeLevel: PrivilegeLevel;
+}
+
```
Comment (anirudh723) [inline src/components/event-details/EventDetails.tsx:23]: This should already be located in `upcoming-events/ducks/types`. Also, like @vrautela mentioned in my PR (that I'm about to merge), we might have some merge conflict here

Comment (anirudh723) [reply src/components/event-details/EventDetails.tsx:23]: And by "here" I meant where the component is declared

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+const { Text } = Typography;
+
+interface EventRegistrationModalProps {
+  eventId: number;
+  eventTitle: string;
+  privilegeLevel: PrivilegeLevel;
+  showEventRegistrationModal: boolean;
+  onCloseEventRegistrationModal: () => void;
+}
```
Comment (anirudh723) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:24]: It's good to put the interfaces/types in your `ducks/types` folder

File: src/components/modals/event-registration-modal/EventRegistrationModal.tsx [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+  const [quantity, setQuantity] = React.useState<number>(1);
+  const [registrationRequest, setRegistrationRequest] = React.useState<
+    AsyncRequest<void, any>
+  >(AsyncRequestNotStarted());
+
+  const updateQuantity = (newValue: string | number | undefined) => {
+    if (typeof newValue === 'number') {
+      setQuantity(newValue);
+    } else {
+      setQuantity(1);
```
Comment (anirudh723) [inline src/components/modals/event-registration-modal/EventRegistrationModal.tsx:75]: Maybe have a constant for initial quantity to avoid magic numbers

**Comment by anirudh723** [comment]
Just merged in my PR, so you can now fix merge conflicts with the most up-to-date master code

**Review by anirudh723 — APPROVED** [review]
Looks great overall! Some small suggestions but otherwise ready to 🚢
